File: packages/ios-client/Sources/FishjamClient/webrtc/helpers/BitrateLimiter.swift
import WebRTC

struct BitrateLimiter {
    static func calculateBitrates(for encodings: [RTCRtpEncodingParameters], maxBitrate: TrackBandwidthLimit) -> [RTCRtpEncodingParameters] {
        switch maxBitrate {
        case .BandwidthLimit(let limit):
            return calculateUniformBitrates(for: encodings, bitrate: limit)
        case .SimulcastBandwidthLimit(let limits):
            return calculateSimulcastBitrates(for: encodings, limits: limits)
        }
    }
    
    private static func calculateSimulcastBitrates(
        for encodings: [RTCRtpEncodingParameters],
        limits: [String: Int]
    ) -> [RTCRtpEncodingParameters] {
        return encodings.map { encoding in
            let encodingLimit = limits[encoding.rid ?? ""] ?? 0
            return encoding.withBitrate(kbps: encodingLimit)
        }
    }
    
    private static func calculateUniformBitrates(
        for encodings: [RTCRtpEncodingParameters],
        bitrate: Int
    ) -> [RTCRtpEncodingParameters] {
        guard !encodings.isEmpty else { return [] }
        guard bitrate != 0 else {
            return encodings.map { $0.withBitrate(kbps: nil) }
        }
        
        // Find minimum scale resolution
        let k0 = encodings.map { $0.scaleResolutionDownByDouble }
                         .min() ?? 1.0
                         
        let bitrateParts = encodings.reduce(0.0) { acc, encoding in
            acc + pow((k0 / encoding.scaleResolutionDownByDouble), 2)
        }
        
        let multiplier = Double(bitrate) / bitrateParts
        
        // Calculate new bitrates
        return encodings.map { encoding in
            let calculatedBitrate = Int(
                multiplier * pow(k0 / encoding.scaleResolutionDownByDouble, 2)
            )
            return encoding.withBitrate(kbps: calculatedBitrate)
        }
    }
}

File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/webrtc/helpers/BitrateLimiter.kt
package com.fishjamcloud.client.webrtc.helpers

import com.fishjamcloud.client.models.TrackBandwidthLimit
import org.webrtc.RtpParameters
import kotlin.math.pow

class BitrateLimiter {
  companion object {
    fun calculateBitrates(
      encodings: List<RtpParameters.Encoding>,
      maxBitrate: TrackBandwidthLimit
    ): List<RtpParameters.Encoding> =
      when (maxBitrate) {
        is TrackBandwidthLimit.BandwidthLimit ->
          calculateUniformBitrates(encodings, maxBitrate.limit)
        is TrackBandwidthLimit.SimulcastBandwidthLimit ->
          calculateSimulcastBitrates(encodings, maxBitrate.limit)
      }

    private fun calculateSimulcastBitrates(
      encodings: List<RtpParameters.Encoding>,
      limits: Map<String, TrackBandwidthLimit.BandwidthLimit>
    ): List<RtpParameters.Encoding> =
      encodings.map { encoding ->
        val encodingLimit = limits[encoding.rid]?.limit ?: 0
        encoding.withBitrateInBps(encodingLimit)
      }

    private fun calculateUniformBitrates(
      encodings: List<RtpParameters.Encoding>,
      bitrate: Int
    ): List<RtpParameters.Encoding> {
      if (encodings.isEmpty()) return emptyList()
      if (bitrate == 0) {
        return encodings.map { it.withBitrateInBps(null) }
      }

      val k0 = encodings.minByOrNull { it.scaleResolutionDownBy ?: 1.0 }

      val bitrateParts =
        encodings.sumOf {
          ((k0?.scaleResolutionDownBy ?: 1.0) / (it.scaleResolutionDownBy ?: 1.0)).pow(2)
        }

      val x = bitrate / bitrateParts

      return encodings.map { encoding ->
        val calculatedBitrate =
          (x * ((k0?.scaleResolutionDownBy ?: 1.0) / (encoding.scaleResolutionDownBy ?: 1.0)).pow(2) * 1024).toInt()
        encoding.withBitrateInBps(calculatedBitrate)
      }
    }
  }
}

private fun RtpParameters.Encoding.withBitrateInBps(bps: Int?): RtpParameters.Encoding {
  val encoding =
    RtpParameters.Encoding(
      rid,
      active,
      scaleResolutionDownBy
    )
  encoding.maxBitrateBps = if (bps == 0) null else bps
  return encoding
}
